File: 187-RepeatedDnaSequences/187-RepeatedDnaSequences.cpp
// Last updated: 13/07/2026, 22:24:22
#include <string_view>

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        vector<string> result;
        if (s.size() <= 10) return result;

        unordered_map<string_view, int> freq;
        for (size_t i = 0; i <= s.length() - 10; ++i) {
            string_view window(s.data() + i, 10);
            freq[window]++;
        }

        for (const auto&[s, v] : freq){
            if (v >= 2){
                result.emplace_back(s);
            }
        }
        return result;
    }
};

File: 187-RepeatedDnaSequences/187-RepeatedDnaSequences.rs
// Last updated: 13/07/2026, 22:24:41
use std::collections::HashMap;

impl Solution {
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
        if s.len() < 10 {
            return Vec::new();
        }

        let mut result: Vec<String> = Vec::new();
        let mut freq: HashMap<&[u8], u32> = HashMap::new();

        for window in s.as_bytes().windows(10) {
            *freq.entry(window).or_insert(0) += 1;
        }

        for (bytes, &val) in freq.iter(){
            if (val >= 2){
                result.push(String::from_utf8(bytes.to_vec()).unwrap());
            }
        }
        return result;
    }
}